### ISSUE TITLE
Vim9: E1036 might unexpectedly shadow E1012 when using popup_getoptions() in assignment

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#7160]
**Describe the bug**

In Vim9 script, `E1036` might unexpectedly shadow `E1012` when using `popup_getoptions()` in an assignment.

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9script
        def Func()
            var menu_winid = popup_menu('', {})
            var opts = #{ key: '' }
            opts = popup_getoptions(menu_winid)
            eval opts.line - 1
        enddef
        Func()
    EOF
    )

`E1036` is raised:

    E1036: - requires number or float arguments

**Expected behavior**

`E1012` is raised:

    E1012: Type mismatch; expected dict<func(...): any> but got dict<any>

**Environment**

 - Vim version: 8.2 Included patches: 1-1857
 - OS: Ubuntu 16.04.7 LTS
 - Terminal: xterm(360)

**Additional context**

If the last line in the body of `Func()` is removed, `E1012` is correctly raised:
```vim
vim9script
def Func()
    var menu_winid = popup_menu('', {})
    var opts = #{ key: '' }
    opts = popup_getoptions(menu_winid)
enddef
Func()
```
    E1012: Type mismatch; expected dict<func(...): any> but got dict<any>

---

The fact that `E1036` can shadow `E1012` makes debugging harder because it requires us to unravel a cascade of errors; besides, it's inconsistent:
```vim
vim9script
def Func()
    invalid1
    invalid2
enddef
Func()
```
    E476: Invalid command: invalid1

Notice how this time, the first syntax error is reported, not the last one.

---

If `- 1` is replaced with `+ 1`, `E1012` is still shadowed, but this time by another error (`E1051`):
```vim
vim9script
def Func()
    var menu_winid = popup_menu('', {})
    var opts = #{ key: '' }
    opts = popup_getoptions(menu_winid)
    eval opts.line + 1
enddef
Func()
```
    E1051: Wrong argument type for +

---

If `popup_getoptions()` is removed, the issue disappears:
```vim
vim9script
def Func()
    var opts = #{key: ''}
    opts = #{line: 1}
    eval opts.line - 1
enddef
Func()
```
    E1012: Type mismatch; expected dict<string> but got dict<number>

**Comment by lacygoill** [comment]
> E1012 is raised:
>
>     E1012: Type mismatch; expected dict<func(...): any> but got dict<any>

Sorry, I meant:

    E1012: Type mismatch; expected dict<string> but got dict<any>

**Comment by brammool** [comment]
E1036 is given when the function is compiled.  This can be detected at compile time.
E1012 can only be given when the function is executed, since during compilation converting "any" into something else might be OK.
Thus this is working as intended.

**Comment by lacygoill** [comment]
> E1036 is given when the function is compiled. This can be detected at compile time.
> E1012 can only be given when the function is executed

I see, thank you.  The 2 errors happen at different stages.  I should have noticed the difference in the error messages ("processing" for E1012 vs "compiling" for E1036).

> during compilation converting "any" into something else might be OK.

I guess it means that the types `list<any>` and `dict<any>` are accepted instead of more specific types like `list<number>`, `dict<string>`, ... but only at compile time; not at runtime.
```vim
vim9script
def Func()
    var l: list<number> = ['', true]
enddef
defcompile
```
```vim
vim9script
def Func()
    var d: dict<number> = #{a: '', b: true}
enddef
defcompile
```
Vim did report an error at compile time up to [8.2.1691](https://github.com/vim/vim/commit/5e654230777ad21363a929dce3cfe0387da031a7).  But it caused some issues like [this one](https://github.com/vim/vim/issues/6959), and [these ones](https://github.com/vim/vim/issues/6959#issuecomment-693513694).

However, for some reason, `list<any>` and `dict<any>` are not accepted at compile time when they're nested:
```vim
vim9script
def Func()
    var l: list<list<number>> = [['', true]]
enddef
defcompile
```
    E1012: Type mismatch; expected list<list<number>> but got list<list<any>>

**Comment by lacygoill** [comment]
All of this is still a bit confusing:
```vim
vim9script
def Func()
    var l: list<number> = ['', true]
enddef
defcompile
```
    no error
```vim
vim9script
def Func()
    var l: list<number> = ['', true]
enddef
Func()
```
    E1012: Type mismatch; expected list<number> but got list<any>
```vim
vim9script
def Func()
    var l: list<list<number>> = [['', true]]
enddef
defcompile
```
    E1012: Type mismatch; expected list<list<number>> but got list<list<any>>

I'll need to re-read the bug report #6959 and the help.

**Comment by lacygoill** [comment]
Something else is confusing:
```vim
vim9script
def Func()
    var l: list<number> = [0, '']
enddef
defcompile
```
    no error
```vim
vim9script
def Func()
    var l: list<number> = [0]
    add(l, [''])
enddef
defcompile
```
    no error
```vim
vim9script
def Func()
    var l: list<number> = [0]
    l += ['']
enddef
defcompile
```
    E1012: Type mismatch; expected list<number> but got list<string>

If a value with the type `list<any>` is acceptable at compile time, then I would expect the last snippet not to raise any error, just like the first two.

---

Even more confusing:
```vim
vim9script
def Func()
    var l: list<number> = [0]
    add(l, '')
enddef
Func()
```
This snippet doesn't raise any error, neither at compile time, nor at runtime.  And yet, the type of the final value assigned to `l` (`list<any>`) does not match the type specified in the declaration (`list<number>`).

**Comment by brammool** [comment]
> > E1036 is given when the function is compiled. This can be detected at compile time.
> > E1012 can only be given when the function is executed
> 
> I see, thank you.  The 2 errors happen at different stages.  I should have noticed the difference in the error messages ("processing" for E1012 vs "compiling" for E1036).

Yes, I recently added the "compiling" text exactly for this reason.

> > during compilation converting "any" into something else might be OK.
> 
> I guess it means that the types `list<any>` and `dict<any>` are
> accepted instead of more specific types like `list<number>`,
> `dict<string>`, ... but only at compile time; not at runtime.
> ```vim
> vim9script
> def Func()
>     var l: list<number> = ['', true]
> enddef
> defcompile
> ```
> ```vim
> vim9script
> def Func()
>     var d: dict<number> = #{a: '', b: true}
> enddef
> defcompile
> ```
> Vim did report an error at compile time up to
> [8.2.1691](https://github.com/vim/vim/commit/5e654230777ad21363a929dce3cfe0387da031a7).
> But it caused some issues like [this
> one](https://github.com/vim/vim/issues/6959), and [these
> ones](https://github.com/vim/vim/issues/6959#issuecomment-693513694).
> 
> However, for some reason, `list<any>` and `dict<any>` are not accepted
> at compile time when they're nested:
> ```vim
> vim9script
> def Func()
>     var l: list<list<number>> = [['', true]]
> enddef
> defcompile
> ```
>     E1012: Type mismatch; expected list<list<number>> but got list<list<any>>

Here you are using a constant, thus the type is known and will never
match.  Combining a string and a boolean, the common type is "any".

-- 
Momento mori, ergo carpe diem

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by brammool** [comment]
> All of this is still a bit confusing:
> ```vim
> vim9script
> def Func()
>     var l: list<number> = ['', true]
> enddef
> defcompile
> ```
>     no error

This should give an error, I'll see if that can be fixed.

> ```vim
> vim9script
> def Func()
>     var l: list<number> = ['', true]
> enddef
> Func()
> ```
>     E1012: Type mismatch; expected list<number> but got list<any>
> ```vim
> vim9script
> def Func()
>     var l: list<list<number>> = [['', true]]
> enddef
> defcompile
> ```
>     E1012: Type mismatch; expected list<list<number>> but got list<list<any>>
> 
> I'll need to re-read the bug report #6959 and the help.

-- 
From "know your smileys":
 %	Bike accident.  A bit far-fetched, I suppose; although...
             o      _     _         _
     _o     /\_   _ \\o  (_)\__/o  (_)
   _< \_   _>(_) (_)/<_    \_| \   _|/' \/
  (_)>(_) (_)        (_)   (_)    (_)'  _\o_

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lacygoill** [comment]
> Here you are using a constant, thus the type is known and will never
> match.  Combining a string and a boolean, the common type is "any".

Yes, but it was inconsistent with the previous snippet, and I didn't know which one was working as intended, and which one was not.

> This should give an error, I'll see if that can be fixed.

I can confirm that [8.2.1864](https://github.com/vim/vim/releases/tag/v8.2.1864) fixes the issue, which makes most of the previous snippets consistent again.

However, I've found 2 other inconsistent snippets; I think one of them explains why `E1036` still shadows `E1012` in the original snippet.  I'll open seperate issues for them.  Thank you for the quick fix.

**Comment by brammool** [comment]
> Something else is confusing:
> ```vim
> vim9script
> def Func()
>     var l: list<number> = [0, '']
> enddef
> defcompile
> ```
>     no error

This gives an error now.

> ```vim
> vim9script
> def Func()
>     var l: list<number> = [0]
>     add(l, [''])
> enddef
> defcompile
> ```
>     no error

The arguments of builtin functions are not checked (yet).

> ```vim
> vim9script
> def Func()
>     var l: list<number> = [0]
>     l += ['']
> enddef
> defcompile
> ```
>     E1012: Type mismatch; expected list<number> but got list<string>
> 
> If a value with the type `list<any>` is acceptable at compile time,
> then I would expect the last snippet not to raise any error, just like
> the first two.

Both give an error now

> Even more confusing:
> ```vim
> vim9script
> def Func()
>     var l: list<number> = [0]
>     add(l, '')
> enddef
> Func()
> ```
> This snippet doesn't raise any error, neither at compile time, not at
> runtime.  And yet, the final value assigned to `l` (`list<any>`) does
> not match the type specified in the declaration (`list<number>`).

Hmm, that can be a problem.  add() doesn't know about the type of "l".
Probably the best is to inline add().

-- 
I AM THANKFUL...
...for the clothes that fit a little too snug because it
means I have more than enough to eat.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lacygoill** [comment]
> Hmm, that can be a problem.  add() doesn't know about the type of "l".
> Probably the best is to inline add().

This is fixed by [8.2.1865](https://github.com/vim/vim/commit/1dcae59957301b6b19aef49af648715f911a1378).

I think the same issue applies to `insert()` and `extend()`:
```vim
vim9script
def Func()
    var l: list<number>
    insert(l, ['x'])
enddef
Func()
```
    no error
```vim
vim9script
def Func()
    var l: list<number>
    extend(l, ['x'])
enddef
Func()
```
    no error

Let me know if you want me to open separate issues for those.

**Comment by brammool** [comment]
Inlining insert() is not so easy.  I'll think about a generic way to check builtin function arguments at compile time.
